rectools/visuals/metrics_app.py

```diff
@@ -0,0 +1,212 @@
+            Additional arguments from `plotly.graph_objects.Layout`
+
+        Returns
+        -------
+        MetricsApp
+            An instance of `MetricsApp`, providing interactive Jupyter widget for metric visualization.
+
+        Examples
+        --------
+        Create interactive widget
+
+        >>> example_df = pd.DataFrame(
+        ...    {
+        ...        Columns.Model: ["Model1", "Model2", "Model1", "Model2", "Model1", "Model2"],
+        ...        Columns.Split: [0, 0, 1, 1, 2, 2],
+        ...        "prec@10": [0.031, 0.025, 0.027, 0.21, 0.031, 0.033],
+        ...        "recall@10": [0.041, 0.045, 0.055, 0.08, 0.036, 0.021],
+        ...        "novelty@10": [2.6, 11.3, 4.3, 9.8, 3.3, 11.2],
+        ...    })
+        >>> app = MetricsApp.construct(
+        ...    models_metrics=example_df,
+        ...    show_legend=True,
+        ...    auto_display=False,
+        ...    layout_kwargs={"width": 800, "height": 600})
+
+        Get plotly chart from widget state
+
+        >>> fig = app.fig
+        >>> fig = fig.update_layout(
+        ...    title="Metrics: recall@10 vs novelty@10",
+        ...    xaxis_title="novelty@10",
+        ...    yaxis_title="recall@10",
+        ...    legend_title="Models",
+        ... )
+        """
+        app = cls(models_metrics, show_legend, auto_display, layout_kwargs)
+        if auto_display:
+            app.display()
+        return app
+
+    @property
+    def metric_names(self) -> tp.List[str]:
+        """List of metric column names from the `models_metrics`."""
+        non_metric_columns = {Columns.Split, Columns.Model}
+        return [col for col in self.models_metrics.columns if col not in non_metric_columns]
+
+    @property
+    def model_names(self) -> tp.List[str]:
+        """Sorted list of model names from `models_metrics`."""
+        return sorted(self.models_metrics[Columns.Model].unique())
+
+    @property
+    def fold_ids(self) -> tp.List[int]:
+        """Sorted list of fold identifiers from the `models_metrics`."""
+        return sorted(self.models_metrics[Columns.Split].unique())
+
+    @property
+    def n_folds(self) -> int:
+        """Total number of unique folds available in the `models_metrics`."""
+        return len(self.fold_ids)
+
+    def _validate_input_data(self) -> None:
+        if not isinstance(self.models_metrics, pd.DataFrame):
+            raise ValueError("Incorrect input type. `metrics_data` should be a DataFrame")
+        if Columns.Split not in self.models_metrics.columns:
+            raise KeyError("Missing `Split` column in `metrics_data` DataFrame")
+        if Columns.Model not in self.models_metrics.columns:
+            raise KeyError("Missing `Model`` column in `metrics_data` DataFrame")
+        if len(self.models_metrics.columns) < 3:
+            raise KeyError("`metrics_data` DataFrame assumed to have at least one metric column")
+
+    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:
```

Comment:
maybe it also makes sense to add caching for this and the next method

Reply:
Done